File: cpp/day_14_Cheapest Flights Within K Stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int K) {
        vector<vector<pair<int, int>>> graph(n, vector<pair<int, int>>());
        int i, j, u, v, cost, ans=INT_MAX, new_cost, flag=0;
        for(i = 0; i < flights.size(); i++){
            u = flights[i][0];
            v = flights[i][1];
            cost = flights[i][2];
            graph[u].push_back({v, cost});
        }
        
        queue<pair<int, int>> q;
        pair<int, int> t;
        vector<int> visited(n, INT_MAX);
        q.push({src, 0});
        visited[src] = true;
        while(!q.empty() && K >= 0){
            for(i = q.size()-1; i >= 0; i--){
                t = q.front();
                q.pop();
                for(j = 0; j < graph[t.first].size(); j++){
                    v = graph[t.first][j].first;
                    new_cost = t.second + graph[t.first][j].second;
                    if(v == dst){
                        ans = min(ans, new_cost);
                        flag = 1;
                    }
                    if(visited[v] > new_cost){
                        visited[v] = new_cost;
                        q.push({v, new_cost});
                    }
                }
            }
            K--;
        }
        return flag ? ans: -1;
    }
};

File: python3/june/day_14_Cheapest Flights Within K Stops.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, K: int) -> int:
        graph = [ [] for _ in range(n)]
        for flight in flights:
            u = flight[0]
            v = flight[1]
            cost = flight[2]
            graph[u].append((v, cost,))
        
        q = deque()
        q.append((src, 0))
        visited = [float('inf') for _ in range(n)]
        ans = float('inf')
        while (len(q) > 0 and K >= 0):
            s = len(q)
            for _ in range(s):
                u, cost = q.popleft()
                for j in range(len(graph[u])):
                    v = graph[u][j][0]
                    new_cost = cost + graph[u][j][1]
                    
                    if dst == v:
                        ans = min(new_cost, ans)

                    if new_cost < visited[v]:
                        visited[v] = new_cost
                        q.append((v, new_cost))
            K -= 1      
        return ans if ans != float('inf') else -1
